File: src/razor-core/DevilDaggersInfo.Razor.Core.Canvas/WebViewCanvas2d.cs
using DevilDaggersInfo.Razor.Core.Canvas.JSRuntime;

namespace DevilDaggersInfo.Razor.Core.Canvas;

// TODO: Complete implementation.
public class WebViewCanvas2d : WebViewCanvas
{
	public WebViewCanvas2d(string id, WebViewRuntimeWrapper wrapper)
		: base(id, wrapper)
	{
	}

	public async Task SetFillStyleAsync(string fillStyle)
		=> await InvokeAsync("c2d.setFillStyle", fillStyle);

	public async Task SetStrokeStyleAsync(string strokeStyle)
		=> await InvokeAsync("c2d.setStrokeStyle", strokeStyle);

	public async Task SetLineWidthAsync(float lineWidth)
		=> await InvokeAsync("c2d.setLineWidth", lineWidth);

	public async Task ClearRectAsync(float x, float y, float width, float height)
		=> await InvokeAsync("c2d.clearRect", x, y, width, height);

	public async Task BeginPathAsync()
		=> await InvokeAsync("c2d.beginPath");

	public async Task ClosePathAsync()
		=> await InvokeAsync("c2d.closePath");

	public async Task MoveToAsync(float x, float y)
		=> await InvokeAsync("c2d.moveTo", x, y);

	public async Task LineToAsync(float x, float y)
		=> await InvokeAsync("c2d.lineTo", x, y);

	public async Task ArcAsync(float x, float y, float radius, float startAngle, float endAngle, bool anticlockwise = false)
		=> await InvokeAsync("c2d.arc", x, y, radius, startAngle, endAngle, anticlockwise ? 1 : 0);

	public async Task CircleAsync(float x, float y, float radius)
		=> await ArcAsync(x, y, radius, 0.0f, MathF.PI * 2);

	public async Task FillAsync()
		=> await InvokeAsync("c2d.fill");

	public async Task StrokeAsync()
		=> await InvokeAsync("c2d.stroke");
}

File: src/razor-core/DevilDaggersInfo.Razor.Core.Canvas/wwwroot/js/webViewCanvas2d.js
window.c2d = {
	getContext: function (id) {
		var c = document.getElementById(id);
		dc = c.getContext("2d");
		return dc;
	},

	clearRect: function (id, x, y, w, h) {
		var dc = c2d.getContext(id);
		dc.clearRect(x, y, w, h);
	},

	setFillStyle: function (id, fs) {
		var dc = c2d.getContext(id);
		dc.fillStyle = fs;
	},
	setStrokeStyle: function (id, ss) {
		var dc = c2d.getContext(id);
		dc.strokeStyle = ss;
	},
	setLineWidth: function (id, lw) {
		var dc = c2d.getContext(id);
		dc.lineWidth = lw;
	},
	beginPath: function (id) {
		var dc = c2d.getContext(id);
		dc.beginPath();
	},
	closePath: function (id) {
		var dc = c2d.getContext(id);
		dc.closePath();
	},
	moveTo: function (id, x, y) {
		var dc = c2d.getContext(id);
		dc.moveTo(x, y);
	},
	lineTo: function (id, x, y) {
		var dc = c2d.getContext(id);
		dc.lineTo(x, y);
	},
	arc: function (id, x, y, radius, startAngle, endAngle, anticlockwise) {
		var dc = c2d.getContext(id);
		dc.arc(x, y, radius, startAngle, endAngle, anticlockwise !== 0);
	},
	fill: function (di) {
		var dc = c2d.getContext(di);
		dc.fill();
	},
	stroke: function (di) {
		var dc = c2d.getContext(di);
		dc.stroke();
	},
};
